src/workerd/jsg/util.h

```diff
@@ -394,4 +394,18 @@ inline bool isFinite(double value) {
   return !(kj::isNaN(value) || value == kj::inf() || value == -kj::inf());
 }
 
+// ======================================================================================
+
+class Lock;
+
+v8::MaybeLocal<v8::String> newExternalOneByteString(Lock& js, kj::ArrayPtr<const char> buf);
+v8::MaybeLocal<v8::String> newExternalTwoByteString(Lock& js, kj::ArrayPtr<uint16_t> buf);
+// Creates v8 Strings from buffers not on the v8 heap. These do not copy and do not
+// take ownership of the buf. Care must be taken to ensure that the buffers are kept
```

Comment:
This feels error-prone to me. How can the caller possibly know when V8 has garbage-collected the reference? I would be more comfortable with these methods taking `kj::Array` rather than `kj::ArrayPtr`, and disposing the array when V8 collects the external reference.

Reply:
If I understand the suggestion correctly and each `kj::Array` would be a copy, that runs counter to the goal here for builtins. For the built-in JS/TS support, the content of the built-in modules are compiled into the binary, for example...

```
extern const ::kj::byte EMBED_ts_node_events_js_begin;
extern const ::kj::byte EMBED_ts_node_events_js_end;
#define EMBED_ts_node_events_js (::kj::ArrayPtr<const ::kj::byte>( \
    &EMBED_ts_node_events_js_begin, &EMBED_ts_node_events_js_end))
```

For these particular functions, the caller wouldn't care when V8 has garbage-collected the reference because the underlying data is immutable for the lifetime of the process and is never disposed of. If the reference is going to hold a copy of the data in a kj::Array then there's no real point to doing this as opposed to just having the v8 heap hold the data.

Reply:
A `kj::Array` does not need to be a copy, it could use a custom disposer.

However, I had missed that the intent is to use this with builtins. If it's always the case that the code is a static constant that's valid for the lifetime of the process, then I think it's OK. We should explicitly document this requirement, though: the array _must_ point to a static constant with infinite lifetime. I don't want anyone passing in a dynamic string and then trying to guess when it's safe to destroy.